Ensure that GRPC_VERBOSITY is checked before logging

ncteisen opened the pull request:
This fixes https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/grpc-io/3ej42lIE8ZA/COPLM70bCAAJ

There were two points in which data was logged before GRPC_VERBOSITY was checked.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

yang-g commented on src/core/lib/support/env.h:
```diff
@@ -36,6 +36,8 @@ char *gpr_getenv(const char *name);
 /* Sets the the environment with the specified name to the specified value. */
 void gpr_setenv(const char *name, const char *value);
 
+char *gpr_getenv_silent(const char *name, char** dst);
```
Please add comments to this  function.

ncteisen commented:
@ctiller I added a comment explaining the one use case of the silent version and I added it to to core_banned_functions. Does this address the concerns you mentioned to me yesterday?

Even though this solution is a little messy, I think buffering up pre init log calls would add even more complexity and potentially hide some crashes.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_trickle.BM_PumpStreamServerToClient_Trickle_1_4k.opt.new: 1


[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] Performance differences noted:
Benchmark                                      svr_transport_stalls_per_iteration
---------------------------------------------  ------------------------------------
BM_PumpStreamServerToClient_Trickle/128M/128M  +9999%
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

ncteisen commented:
I'm gonna merge this unless there are serious objections, @ctiller. It's not that clean, but it solves a real problem of unwanted, unneeded logs

ncteisen commented:
Tsan: #11482 
Windows: #10807

ncteisen commented:
This fixes #8104.
